Source/Applications/SystemCenter/SystemCenterEngine.cs

```diff
@@ -173,21 +235,35 @@ public void ReloadConfiguration()
         public void ReloadSystemSettings()
         {
             ConfigurationFile configurationFile;
-            CategorizedSettingsElementCollection category;
 
             // Reload the configuration file
             configurationFile = ConfigurationFile.Current;
             configurationFile.Reload();
+
+            // Reconstruct known connection factories
             AdoDataConnection.ReloadConfigurationSettings();
+            foreach (string key in m_connectionFactories.Keys)
+                CreateAndAddFactory(configurationFile, key);
```

Comment:
This is the only place where `AddOrUpdate()` is actually relevant. It can be rewritten as a one-liner using the assignment operator.
```suggestion
                m_connectionFactories[key] = new DatabaseConnectionFactory(configurationFile, key);
```

Therefore, you can completely repurpose the `CreateAndAddFactory()` method to use `GetOrAdd()` instead, to support the other two callsites.